### PR TITLE
Add support for "beatmap not downloaded / downloading" user states

### PR DESCRIPTION
**Pull request by frenzibyte**
- [ ] Depends on ppy/osu#11289

Note that I haven't tested this due to lack of a local dev server (I tried setting one, but I blew up on the way and closed everything).

### CODE REVIEW

File: osu.Server.Spectator/Hubs/MultiplayerHub.cs [hunk 1]
```diff
@@ -515,6 +516,16 @@ private void ensureValidStateSwitch(MultiplayerRoom room, MultiplayerUserState o
         {
             switch (newState)
             {
+                case MultiplayerUserState.BeatmapNotDownloaded:
```
Comment (peppy) [inline osu.Server.Spectator/Hubs/MultiplayerHub.cs:519]: probably just combine with idle below, since they are basically the same thing.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Server.Spectator/Hubs/MultiplayerHub.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Server.Spectator/Hubs/MultiplayerHub.cs:519]: done.

**Comment by bdach** [comment]
I'm not actually sure I'd agree with being able to transition into `BeatmapNotDownloaded` from any state. Surely it doesn't make sense to do so from something like `Playing`. I think the flow should be something like the following:

obraz

I guess there's an argument to be made that permitting any state to transition to `BeatmapNotDownloaded` is more resilient against a possible dropped message about a transition to `Idle`. I'd definitely permit `DownloadingBeatmap` -> `Ready` in that case, however.

**Comment by peppy** [comment]
Yeah, while thinking about this i almost proposed keeping the map download state separate from the main state enum to keep things simple. Still thinking about this one.

**Comment by frenzibyte** [comment]
> I'm not actually sure I'd agree with being able to transition into `BeatmapNotDownloaded` from any state. Surely it doesn't make sense to do so from something like `Playing`. I think the flow should be something like the following:
> 
> obraz
> 
> I guess there's an argument to be made that permitting any state to transition to `BeatmapNotDownloaded` is more resilient against a possible dropped message about a transition to `Idle`. I'd definitely permit `DownloadingBeatmap` -> `Ready` in that case, however.

Was thinking like that as well, previously I had `BeatmapNotDonwloaded` only accept being transitioned from `Ready`, `Idle` and `DownloadingBeatmap`, but I thought I was overcomplicating stuff, and maybe it would potentially be allowed to delete a map from whatever state you're in, for god knows what reason. 

But I see we're set in one direction now, will change as proposed.

**Comment by frenzibyte** [comment]
> Yeah, while thinking about this i almost proposed keeping the map download state separate from the main state enum to keep things simple. Still thinking about this one.

I'll have a go on this one, sounds logical.

**Comment by frenzibyte** [comment]
Closing as I'm going in a different direction, as above.

**Pull request closed without merging**